app/Http/Requests/CreateCircleRequest.php

```diff
@@ -0,0 +1,79 @@
+<?php
+
+namespace App\Http\Requests;
+
+use App\Models\Circle;
+use App\Models\Protocol;
+use App\Models\User;
+use Illuminate\Foundation\Http\FormRequest;
+
+class CreateCircleRequest extends FormRequest
+{
+    /**
+     * Determine if the user is authorized to make this request.
+     *
+     * @return bool
+     */
+    public function authorize()
+    {
+        return true;
+    }
+
+    /**
+     * Get the validation rules that apply to the request.
+     *
+     * @return array
+     */
+    protected function prepareForValidation()
+    {
+        $data = json_decode($this->get('data'), true);
+        $this->merge([
+            'user_name' => !empty($data['user_name']) ? $data['user_name'] : null,
+            'address' => !empty($data['address']) ? strtolower($data['address']) : null,
+            'circle_name' => !empty($data['circle_name']) ? $data['circle_name'] : null
+        ]);
+
+        if (array_key_exists('protocol_id', $data) &&
+            $data['protocol_id']) {
+
+            $this->merge([
+                'protocol_id' => $data['protocol_id']
+            ]);
+        } else if (array_key_exists('protocol_name', $data) &&
+            $data['protocol_name']
+        ) {
+            $this->merge([
+                'protocol_name' => $data['protocol_name']
+            ]);
+        }
+    }
+
+    public function rules()
+    {
+        $addressValidations = ['required', 'string', 'size:42'];
+        if ($this->protocol_id) {
+            $addressValidations[] = function ($attribute, $value, $fail) {
+                $protocol = Protocol::find($this->protocol_id);
+                if ($protocol) {
+                    $circles = Circle::where('protocol_id', $this->protocol_id)->pluck('id')->toArray();
+                    if (count($circles)) {
+                        $exists = User::byAddress($value)->where('role', 1)->whereIn('circle_id', $circles)->exists();
```

Comment:
Could we make an ENUM or constant of the user roles so rather than putting `1` here, we use a constant like `USER:ROLES:ADMIN` or similar?

Reply:
I'm seeing this is used pretty widely so this would be good to change at some point, but not in this PR.

Reply:
updated